clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2081,19 +2448,30 @@ namespace {
                               Cast->getType()->isCheckedPointerPtrType());
       ProofStmtKind Kind = IsStaticPtrCast ? ProofStmtKind::StaticBoundsCast :
                              ProofStmtKind::BoundsDeclaration;
-      ProofResult Result =
-        ProveBoundsDeclValidity(TargetBounds, SrcBounds, Cause, nullptr, Kind);
+      FreeVariableListTy FreeVars;
+      ProofResult Result = ProveBoundsDeclValidity(
+          TargetBounds, SrcBounds, Cause, nullptr, FreeVars, Kind);
       if (Result != ProofResult::True) {
-        unsigned DiagId = (Result == ProofResult::False) ?
-          diag::error_static_cast_bounds_invalid :
-          (CSS != CheckedScopeSpecifier::CSS_Unchecked ?
-           diag::warn_checked_scopestatic_cast_bounds_invalid :
-           diag::warn_static_cast_bounds_invalid);
+        // Which diagnostic message to print?
+        unsigned DiagId =
+            (Result == ProofResult::False)
+                ? (TestFailure(Cause, ProofFailure::HasFreeVariables)
+                       ? diag::error_static_cast_bounds_unprovable
+                       : diag::error_static_cast_bounds_invalid)
```

Comment:
I believe the `error_static_cast_bounds_unprovable` error code should be deleted (delete this as well from DiagnosticSemaKinds.td). If the proof result is false, only the existing `errror_static_cast_bounds_invalid` message should be emitted.